config/thisroot.sh

```diff
@@ -238,11 +238,27 @@ if [ -z "${SOURCE}" ]; then
       return 1
    fi
 else
-   # get param to "."
-   thisroot="$(dirname "${SOURCE}")"
-   ROOTSYS=$(cd "${thisroot}/.." > /dev/null && pwd); export ROOTSYS
-   if [ -z "$ROOTSYS" ]; then
-      echo "ERROR: \"cd ${thisroot}/..\" or \"pwd\" failed" >&2
+   if [ -L ${SOURCE} ] ; then
+      if [ -e ${SOURCE} ] ; then
+         SOURCE=`readlink "$SOURCE"`
+      else
+         echo "ERROR: Broken link ${SOURCE}"
+         ROOTSYS=; export ROOTSYS
+         return 1
+      fi
+   fi
+   if [ -f ${SOURCE} ] ; then
+      # get param to "."
+      thisroot="$(dirname "${SOURCE}")"
+      
+      ROOTSYS=$(cd "${thisroot}/.." > /dev/null && pwd); export ROOTSYS
+      if [ -z "$ROOTSYS" ]; then
+         echo "ERROR: \"cd ${thisroot}/..\" or \"pwd\" failed" >&2
+         return 1
+      fi
+   else
+      echo "Missing file ${SOURCE}"
+      ROOTSYS=; export ROOTSYS
```

Comment:
I think using the canonicalise function of readlink should make this very easy. We just need to take away `bin/` afterwards
```suggestion
   thisrootdir=$(dirname $(readlink -e "${SOURCE}"))                                                                          
   export ROOTSYS=${thisrootdir%/*}
```

Reply:
Good questions ...
If the link is broken, there is no script to source I believe. So I don't think that this case needs to be considered.

If the folder does not exist, I believe there is no script to source either, so I guess we don't need to do anything.

I didn't look up the syntax for other shells, because this is the `.sh` variant.